### PR TITLE
Use mutator extension for controls_if block

### PR DESCRIPTION
**Pull request by rachel-fenichel**
Still needs tests.

### CODE REVIEW

File: core/extensions.js [hunk 1]
```diff
@@ -77,6 +77,60 @@ Blockly.Extensions.registerMixin = function(name, mixinObj) {
...
+    }
+
+    if (this.hasOwnProperty('mutationToDom')) {
+      throw new Error('Error: tried to apply mutation "' + name +
+          '" to a block that already has a domToMutation function.  Block id: ' +
+          this.id);
+    }
+
+    // Possibly add a mutator dialog.
+    if (opt_blockList) {
```
Comment (RoboErikG) [inline core/extensions.js]: You should also check if the compose and decompose properties exist and if so create the Mutator dialog.

File: core/block.js [hunk 2]
```diff
@@ -985,6 +985,11 @@ Blockly.Block.prototype.appendDummyInput = function(opt_name) {
  * @param {!Object} json Structured data describing the block.
  */
 Blockly.Block.prototype.jsonInit = function(json) {
+  // Add the mutator to the block
+  if (json['mutator'] !== undefined) {
```
Comment (AnmAtAnm) [inline core/block.js]: Push this down to just before the other extensions, so all inputs, etc are initialized before this function is called.

File: core/block.js [hunk 3]
```diff
@@ -985,6 +985,11 @@ Blockly.Block.prototype.appendDummyInput = function(opt_name) {
...
+    Blockly.Extensions.apply(json['mutator'], this);
+  }
```
Comment (AnmAtAnm) [inline core/block.js]: Check domToMutation/mutationtoDom exist after apply. The extension may not be defined by registerMutator().

Also check those didn't change after each json.extensions entry (line 1051).

Comment (rachel-fenichel) [inline core/block.js]: Should I instead have those checks in the extensions file?

File: core/block.js (hunk 2 above)
Comment (rachel-fenichel) [inline core/block.js]: Okay.  I misread your initialization order in your email.

**Comment by rachel-fenichel** [comment]
I looked at the math mutator, and this code isn't sufficient.  That mutator combines a mixin and a function, and the it doesn't make sense to split them up:

```
Blockly.Constants.Math.IS_DIVISIBLE_MUTATOR_EXTENSION = function() {
  this.mixin(Blockly.Constants.Math.IS_DIVISIBLEBY_MUTATOR_MIXIN);
  this.getField('PROPERTY').setValidator(function(option) {
    var divisorInput = (option == 'DIVISIBLE_BY');
    this.sourceBlock_.updateShape_(divisorInput);
  });
};
```

**Comment by RoboErikG** [comment]
Is there anything wrong with a mutator doing additional work besides just adding the mixin? It seems you'd just use Extensions.register(Blockly.Constants.Math.IS_DIVISIBLE_MUTATOR_EXTENSION) instead of registerMutator.

**Comment by RoboErikG** [comment]
Looks reasonable

File: core/extensions.js [hunk 4]
```diff
@@ -135,14 +121,160 @@ Blockly.Extensions.registerMutator = function(name, mixinObj, opt_blockList) {
...
+/**
+ * Check that the current mutator properties match a list of old mutator
+ * properties.  This should be called after applying a non-mutator extension,
+ * to verify that the extension didn't change properties it shouldn't.
+ * @param {!Array.<Object>} oldProperties The old values to compare to.
+ * @param {!Blockly.Block} block The block to inspect for new values.
+ * @return {boolean} True if the property lists match.
+ */
+Blockly.Extensions.mutatorPropertiesMatch = function(oldProperties, block) {
+  var match = true;
```
Comment (AnmAtAnm) [inline core/extensions.js]: Alternatively, you can use an early return false in both cases, instead of a var.

**Comment by rachel-fenichel** [comment]
Added tests.  PTAL.

**Comment by AnmAtAnm** [comment]
Quite thorough on the tests.  LGTM

**Pull request merged**